internal/jobcontainers/logon.go

```diff
@@ -1,35 +1,133 @@
 package jobcontainers
 
 import (
+	"context"
 	"fmt"
 	"strings"
+	"unsafe"
 
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/log"
 	"github.com/Microsoft/hcsshim/internal/winapi"
 	"github.com/pkg/errors"
 	"golang.org/x/sys/windows"
 )
 
-// processToken returns a user token for the user specified by `user`. This should be in the form
-// of either a DOMAIN\username or just username.
-func processToken(user string) (windows.Token, error) {
+func randomPswd() (*uint16, error) {
+	g, err := guid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+	return windows.UTF16PtrFromString(g.String())
+}
+
+func groupExists(groupName string) bool {
+	var p *byte
+	if err := winapi.NetLocalGroupGetInfo(
+		"",
+		groupName,
+		1,
+		&p,
+	); err != nil {
+		return false
+	}
+	defer windows.NetApiBufferFree(p)
+	return true
+}
+
+// makeLocalAccount creates a local account with the passed in username and a randomly generated password.
+// If `groupName` is not an empty string the user will be added to group if it exists.
+func makeLocalAccount(ctx context.Context, user, groupName string) (*uint16, error) {
+	// Create a local account with a random password
+	pswd, err := randomPswd()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate random password: %w", err)
+	}
+
+	userUTF16, err := windows.UTF16PtrFromString(user)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encode username to UTF16: %w", err)
+	}
+
+	usr1 := &winapi.UserInfo1{
+		Name:     userUTF16,
+		Password: pswd,
+		Priv:     winapi.USER_PRIV_USER,
+		Flags:    winapi.UF_NORMAL_ACCOUNT | winapi.UF_DONT_EXPIRE_PASSWD,
+	}
+	if err := winapi.NetUserAdd(
+		"",
+		1,
+		(*byte)(unsafe.Pointer(usr1)),
+		nil,
+	); err != nil {
+		return nil, fmt.Errorf("failed to create user %s: %w", user, err)
+	}
+
+	log.G(ctx).WithField("username", user).Debug("Created local user account for job container")
+
+	sid, _, _, err := windows.LookupSID("", user)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup SID for user %q: %w", user, err)
+	}
+
+	sids := []winapi.LocalGroupMembersInfo0{{Sid: sid}}
+	if err := winapi.NetLocalGroupAddMembers(
```

Comment:
what does this do if the groupname is an empty string?

Reply:
If we entered this function we have a valid groupname (which I'd hope Windows doesn't allow an empty string to be a valid group name 😆). Do you still want a guard on top to check if it's empty?

Reply:
It's just that the comment says "If `groupName` is not an empty string the user will be added to group if it exists." 

We could either change the comment to say that this is only called if groupName isn't empty or guard this call

Reply:
Oh great point, I'll change the comment